processing/src/main/java/org/apache/druid/query/filter/TypedInFilter.java

```diff
@@ -541,8 +545,14 @@ private static DruidObjectPredicate<String> createStringPredicate(
         return DruidPredicateMatch.of(index >= 0);
       };
     }
+    final Set<String> stringSet;
+    // when matching strings to numeric match values, use numeric comparator to implicitly cast the string to number
+    if (matchValueType.isNumeric()) {
+      stringSet = new ObjectAVLTreeSet<>(StringComparators.NUMERIC);
```

Comment:
I feel like regardless of that the values arrive as `sortedValues` ; the build cost of this type of `TreeSet` will still be `O(NlogN)` ; and I guess lookup will also be `O(logN)`  

the old approach of a hash lookup `O(N)` ; lookup `O(1)`

I feel like instead of delaying the normalization until the comparision is a bit too much - why not introduce normalization of the input values instead?

Reply:
See the other comments about trying to have consistent behavior no matter how `sqlUseBoundAndSelectors` is set. If `sqlUseBoundAndSelectors` is set to `true` (which was always the old behavior before these new filters and the context flag were introduced), something like `stringCol in (1.0, 2.0, 3.0)` would be planned into an `OrFilter` of `BoundFilter`, e.g. `(1.0 <= stringCol <= 1.0) OR (2.0 <= stringCol <= 2.0) OR (3.0 <= stringCol <= 3.0)`, which is even less efficient than this sorted set.

The goal here is to get this to behave as if though `stringCol` was cast to double, which means we cannot convert the match value set to Strings and use a plain set. We either need to use this numeric comparator, or actually cast the entire string column to a double externally so that this filter uses a double predicate instead of string predicate (not eating the cast in SQL, in which case this filter would instead be using a double predicate because the column would be a double virtual column instead of string regular column).

Since using this comparator is basically guaranteed not to be worse than the older behavior (and in fact slightly better since we don't have to do the more expensive `BoundFilter` check and can use the set instead of the loop that `OR` would need to do), this is the 'easier' approach.

I agree its potentially not the best approach, but the possibly best approach requires quite a lot more work and validation, which is what I really meant by saving not eating the cast for a follow-up PR.